cloudify_tf/terraform/opa.py

```diff
@@ -0,0 +1,246 @@
+            message = '\n'.join(self._validation_errors)
+            raise OPAException(
+                'Validation failed. Reasons: {message}.'.format(
+                    message=message))
+
+    @property
+    def flags(self):
+        if not self._flags:
+            self._flags = self._format_flags(self._flags_from_props)
+        return self._flags
+
+    @staticmethod
+    def from_ctx(_ctx, opa_config=None):
+        opa_config = opa_config or get_opa_config(
+            _ctx.node.properties, _ctx.instance.runtime_properties)
+        _ctx.logger.debug('Using opa_config {}'.format(opa_config))
+        return OPA(
+            _ctx.logger,
+            _ctx.deployment.id,
+            _ctx.instance.id,
+            **opa_config)
+
+    @contextmanager
+    def configfile(self):
+        with NamedTemporaryFile(mode="w+", delete=False) as f:
+            if self.config:
+                json.dump(self.config, f)
+                f.flush()
+                shutil.move(f.name, self.terraform_root_module+'/config.json')
+                try:
+                    yield 'config.json'
+                except Exception:
+                    raise
+            else:
+                try:
+                    yield
+                except Exception:
+                    raise
+
+    # Evaluate OPA policies against a given input file, such as a Terraform plan
+    # The decision specified by the decision kwarg is evaluated, and the JSON
+    ## evaluation result is returned
+    def evaluate_policy(self, input_file=None, decision="deny", command_extension=None):
+
+        if not input_file:
+            raise OPAException("No input file defined for OPA to execute against.")
+
+        with self.configfile() as config_file:
+            basic_commands = ['exec']
+
+            basic_commands.extend(['--decision', decision])
+
+            # A user may specify a single policy bundle path or multiple
+            ## Figure out which and append CLI args as needed.
+            if isinstance(self.policy_bundle, str):
+                basic_commands.extend(['--bundle', self.policy_bundle])
+            elif isinstance(self.policy_bundle, list):
+                for bundle in self.policy_bundle:
+                    basic_commands.extend(['--bundle', bundle])
+            else:
+                raise OPAException("Unable to build OPA command string. " +
+                "Bundle must be string or list, got {}".format(type(self.policy_bundle)))
+
+            if config_file:
+                basic_commands.extend(['--config-file', config_file])
+
+            if command_extension:
+                basic_commands.extend(command_extension)
+
+            basic_commands.append(input_file)
+
+            command = self.merged_args(self.flags, basic_commands)
+
+            command.insert(0, self.executable_path)
+            return self.execute(command, self.terraform_root_module, self.env,
+                                return_output=True)
```

Comment:
```suggestion
            result = self.execute(command, self.terraform_root_module, self.env, return_output=True)
            return result, json.loads(tf.check_opa())
```
See: https://github.com/cloudify-cosmo/cloudify-terraform-plugin/blob/master/cloudify_tf/terraform/infracost.py#L231

Reply:
I see what you are saying here.

In this case, would it be acceptable to return a boolean for result (indicating whether or not the evaluation passed the policy) and then the JSON result which would be stored as the runtime property?

Reply:
Since there are no runtime properties inside of the `Opa` class, no, but you could store it as in attribute and then reference it later, e.g.:

```
            self._raw_result = self.execute(command, self.terraform_root_module, self.env, return_output=True)

            @property
            def json_result(self):
                return json.loads(self._raw_result)
```

And then in tasks.py, you could do:
```
    ctx.instance.runtime_properties['opa_evaluation_result'] = tf.opa.json_result
```

Reply:
I've implemented this similar to what is done with infracost. Let me know if it needs further revision.